### DISCUSSION TITLE
Issues with custom decorators

### DISCUSSION THREAD

**Discussion by ThirumalaiK** [fastapi/fastapi#11187] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code
main.py
----------
```python

from __future__ import annotations

from functools import wraps

from fastapi import FastAPI, Request, Response

from auth import auth_external

app = FastAPI()


def auth_internal(fn):
    @wraps(fn)
    async def async_wrapper(*args, **kwargs):
        return await fn(*args, **kwargs)

    return async_wrapper


@app.get("/{_:path}")
@auth_external
async def route1(_: Request):
    return Response(status_code=204)

```

auth.py
----------
```python
from __future__ import annotations

from functools import wraps


def auth_external(fn):
    @wraps(fn)
    async def async_wrapper(*args, **kwargs):
        return await fn(*args, **kwargs)

    return async_wrapper
```

### Description

I've experienced weird behaviour with decorators from defined on other modules. 

Getting `pydantic.errors.PydanticUndefinedAnnotation: name 'Request' is not defined` on using a decorator defined on other modules (i.e. `auth.auth_external`). But, the same works when the decorator copied to the main module (i.e `auth_internal`)

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

Python 3.10.12

**Comment by YuriiMotov** [comment]
Can confirm that behavior.
But everything works fine if you either remove `from __future__ import annotations` from main.py or add `from fastapi import Request` to auth.py

**Comment by pfan94** [comment]
`@wraps(fn, updated=("__dict__", "__globals__"))` worked for me

**Comment by ghost** [comment]
This is a good fix. However, in some cases, you cannot change the decorator's behavior (e.g. a third-party library). This is how we worked around it: https://github.com/fastapi/fastapi/discussions/12496

I'll see if your fix is applicable in our case, as that would be preferable

**Comment by YuriiMotov** [comment]
Try using this approach:

```
from __future__ import annotations

from functools import wraps
from typing import get_type_hints

from fastapi import FastAPI, Request, Response

from auth import auth_external

app = FastAPI()

def evaluate_annotations(fn):
    fn.__annotations__ = get_type_hints(fn)
    return fn


@app.get("/internal/{_:path}")
@auth_internal
async def route_internal(_: Request):
    return Response(status_code=204)


@app.get("/external/{_:path}")
@auth_external
@evaluate_annotations
async def route_external(_: Request):
    return Response(status_code=204)

```

When `__future__.annotations` is imported, it forces Python to store annotations as strings in runtime. And this brakes the logic.
So, I added `evaluate_annotations` decorator that just evaluates annotations

**Comment by YuriiMotov** [comment]
This was fixed by https://github.com/fastapi/fastapi/pull/5077, works well with FastAPI 0.123.5+
